pkg/controllers/machine/lifecycle/launch_timeout.go

```diff
@@ -0,0 +1,60 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/metrics"
+)
+
+type LaunchTimeout struct {
+	clock      clock.Clock
+	kubeClient client.Client
+}
+
+// launchTimeout is a heuristic time that we expect to succeed with our cloudprovider.Create() call
+// If we don't succeed within this time, then we should delete and try again through some other mechanism
+const launchTimeout = time.Minute * 2
```

Comment:
Maybe worth considering a heuristic that models making progress, as opposed to time. Another thought -- can we classify retriable vs nonretriable errors?

Reply:
How do you mean? Also, we have kind-of already classified retriable vs. non-retriable errors. I'm concerned about not doing a self-delete after some amount of time though; I worry that we won't catch every non-retriable error and then get in a situation where a machine never launches and we are never able to re-schedule. I think this keeps that from ever happening by being a catch-all for errors that we think are retriable but might not ever succeed

Reply:
We fail loudly in the case of `InsufficientCapacity` so that's one "non-retriable" error that we special case. Only other one that I'm aware of is a configuration error like an AMI mismatch. I'm thinking this one could also fall into the "InsufficentCapacity" bucket so that we consider anything that results in us not having any instance types available (for whatever reason) is in this bucket

Reply:
In cases where there is a mis-configuration, though, I worry about immediately terminating, particularly because this might get really spammy against the api-server if there's a launch, failed AMI discovery, then a delete, then a re-schedule over and over again